src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1084 @@
+                        if (CanSkip(rewrittenPattern))
+                        {
+                            continue;
+                        }
+
+                        BoundPattern rewrittenRecursive = PullBinaryPatternsOut(rewrittenPattern, inputType, makeDeconstructionPattern, args: (node, deconstruction[i], discards, i));
+                        builder.Add(rewrittenRecursive);
+
+                        if (!_negated)
+                        {
+                            inputType = rewrittenRecursive.NarrowedType;
+                        }
+                    }
+                }
+
+                if (!node.Properties.IsDefault)
+                {
+                    foreach (BoundPropertySubpattern property in node.Properties)
+                    {
+                        BoundPattern rewrittenPattern = VisitWithInputTypeFromPattern(property.Pattern);
+
+                        if (CanSkip(rewrittenPattern))
+                        {
+                            continue;
+                        }
+
+                        BoundPattern rewrittenRecursive = PullBinaryPatternsOut(rewrittenPattern, inputType, makePropertyPattern, args: (node, property));
+                        builder.Add(rewrittenRecursive);
+
+                        if (!_negated)
+                        {
+                            inputType = rewrittenRecursive.NarrowedType;
+                        }
+                    }
+                }
+
+                BoundPattern result = MakeBinary(disjunction: _negated, builder: builder, syntax: node.Syntax);
+                builder.Free();
+                return result;
+
+                // Make `Type ( ..., _, <newPattern>, _, ... )`
+                static BoundPattern makeDeconstructionPattern(BoundPattern newPattern, TypeSymbol inputType,
+                    (BoundRecursivePattern recursiveTemplate, BoundPositionalSubpattern subpatternTemplate, ImmutableArray<BoundPositionalSubpattern> discards, int i) args)
+                {
+                    newPattern = WithInputTypeCheckIfNeeded(newPattern, args.subpatternTemplate.Pattern.InputType);
+                    ImmutableArray<BoundPositionalSubpattern> newSubpatterns = args.discards.SetItem(args.i, args.subpatternTemplate.WithPattern(newPattern));
+
+                    BoundPattern newRecursive = new BoundRecursivePattern(
+                        newPattern.Syntax, declaredType: args.recursiveTemplate.DeclaredType, deconstructMethod: args.recursiveTemplate.DeconstructMethod,
+                        deconstruction: newSubpatterns,
+                        properties: default, isExplicitNotNullTest: false, variable: null, variableAccess: null,
+                        args.recursiveTemplate.InputType, args.recursiveTemplate.NarrowedType, args.recursiveTemplate.HasErrors);
+
+                    newRecursive = WithInputTypeCheckIfNeeded(newRecursive, inputType);
```

Comment:
>newRecursive = WithInputTypeCheckIfNeeded(newRecursive, inputType); [](http://example.com/codeflow?start=20&length=67)

It feels like it would be better to let ```MakeBinary``` deal with adjusting the input types when it combines the patterns.